### PR TITLE
WIP: Initial import of the alloc api (testing only in str.c)

### PR DESCRIPTION
**Pull request by radare**

### CODE REVIEW

**Comment by radare** [comment]
strdup() is a problem too

File: libr/util/alloc.c [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+R_API void r_alloc_init () {
+	r_alloc_hooks (malloc, calloc, realloc, free);
+}
+
+R_API void *r_malloc(size_t sz) {
+	return _r_malloc? _r_malloc (sz): malloc (sz);
+}
+
+R_API void *r_calloc(size_t sz) {
+	return _r_malloc? _r_malloc (sz): malloc (sz);
```
Comment (ret2libc) [inline libr/util/alloc.c]: ? shouldn't you use `r_calloc`?

File: libr/util/alloc.c [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+R_API void *r_realloc(void *p, size_t sz) {
+	if (_r_realloc) {
```
Comment (ret2libc) [inline libr/util/alloc.c]: very very minor stuff, but... why the `if` here and the ternary operator everywhere else?

File: libr/include/r_util/r_alloc.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+R_API void r_alloc_hooks(RMalloc m, RCalloc c, RRealloc r, RFree f);
+
+#if R_MALLOC_GLOBAL
+R_API RMalloc *r_malloc;
+R_API RCalloc *r_calloc;
+R_API RRealloc *r_realloc;
+R_API RFree *r_free;
+#else
+R_API void *r_malloc(size_t sz);
+R_API void *r_calloc(size_t sz);
```
Comment (ret2libc) [inline libr/include/r_util/r_alloc.h]: wrong signature?

File: libr/util/alloc.c [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+#if R_MALLOC_GLOBAL
+R_API void r_alloc_hooks(RMalloc m, RCalloc c, RRealloc r, RFree f) {
+	r_malloc = m? m: malloc;
+	r_calloc = c? c: calloc;
+	r_realloc = r? r: realloc;
+	r_free = f? f: free;
+}
+
+#else
+static RMalloc *_r_malloc = malloc;
```
Comment (ret2libc) [inline libr/util/alloc.c]: can't you just reuse the global pointers that you use when `R_MALLOC_GLOBAL`?

**Comment by radare** [comment]
> On 2 Oct 2018, at 09:55, Riccardo Schirone <notifications@github.com> wrote:
> 
> @ret2libc commented on this pull request.
> 
> I don't understand what's the big difference between R_MALLOC_GLOBAL == 0 and == 1. The things seem pretty similar to me.
> 
MallocGlobal == 1:

Call function pointers. Which fails to build and is how capstone solves this

MallocGlobal == 0:

Calls a function that checks for those symbols as in a wrap

**Comment by radare** [comment]
Yes. But i want to keep them hidden from the global view

**Comment by ret2libc** [comment]
And is it really that important to have the distinction between MallocGlobal==1/0? I don't see a lot of differences and I'm not sure about the advantages of having both of them.

File: libr/include/r_util/r_alloc.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+R_API void *r_realloc(void *p, size_t sz);
+R_API void r_free(void *p);
+#endif
+
+#else
+
+#define r_malloc(x) malloc((x))
+#define r_calloc(x,y) calloc((x),(y))
```
Comment (ret2libc) [inline libr/include/r_util/r_alloc.h]: one thing i hate in the `R_NEW` macros is that i always forget what is nmemb and what is sz. Could you use more meaningful names for the arguments?

**Comment by radare** [comment]
Its for testing and discussing. Obv not to merge

**Comment by radare** [comment]
Ypu can also commit whatever you want in this pr. This is for discussion and playground

**Comment by ret2libc** [comment]
> Its for testing and discussing. Obv not to merge

perfect. One thing that would help to understand the usefulness of this is to have the alternative implementation of malloc/calloc/etc.

Maybe one alternative implementation may be one that exit with an error as soon as malloc/calloc/etc. returns NULL. It may be used for debugging or similar...

**Comment by radare** [comment]
To use different heap than the app using r2api itself

Choose different alloc impl for benchmarking

Allow to run in kernel, bootloader...

**Comment by XVilka** [comment]
I think it is even possible to use dmalloc http://dmalloc.com/
I used it in one of my embedded projects, it turned up to be VERY useful to spot the bugs.

File: libr/util/alloc.c (hunk 4 above)
Comment (radare) [reply libr/util/alloc.c]: uh?

File: libr/include/r_util/r_alloc.h (hunk 5 above)
Comment (radare) [reply libr/include/r_util/r_alloc.h]: what? swapping nmemb and sz results in the same results,  in fact it's documented, that it doesnt matter at all if you swap them. so i dont see that important at all

**Comment by ret2libc** [comment]
I'm perfectly fine with having this PR in. But before merging it requires some fixes in the actual implementation, because I think there are some unnecessary things and some bugs (wrong calloc signatures, etc.) as noticed in comments.

**Comment by radare** [comment]
yep. i have to address those comments, but the main problem i see is that not everything will be allocated via r_malloc, so we can have problems when using alternative implementations, this is just the base not the whole transition

File: libr/util/alloc.c (hunk 1 above)
Comment (radare) [reply libr/util/alloc.c]: yes

File: libr/util/alloc.c (hunk 4 above)
Comment (radare) [reply libr/util/alloc.c]: maybe we should just assert because there's no sense in  just setting some o f thoe pointers and not others

**Comment by radare** [comment]
pls review again

**Comment by radare** [comment]
if we use r_malloc we can do checks like if (allocSize > MB(16)) { ...

**Comment by radare** [comment]
@ret2libc what do you think about using static inline instead? actually we shuold be able to configure the malloc-wrapper-level. but im still unsure about the benefits:

pros:
* use different allocation engine
* detect who tries to do HUGE allocations
* ability to implement our memory profiler/logs
* ability to run in kernel/grub by changing the malloc callbacks

cons:
* we should change zillions of calls to malloc/calloc/free
* easy that any other plugin uses libc apis inatead of r_alloc ones and will crash
* require global vars which are evil

**Pull request merged**